### ISSUE TITLE
Tracking Issue for RFC 3137: let-else statements

### ISSUE THREAD

**Issue by joshtriplett** [rust-lang/rust#87335]
This is a tracking issue for [RFC 3137: let-else statements](https://github.com/rust-lang/rfcs/pull/3137).
The feature gate for the issue is `#![feature(let_else)]`.

### About tracking issues

Tracking issues are used to record the overall progress of implementation.
They are also used as hubs connecting to other relevant issues, e.g., bugs or open design questions.
A tracking issue is however *not* meant for large scale discussion, questions, or bug reports about a feature.
Instead, open a dedicated issue for the specific matter and add the relevant feature gate label.

### Steps

- [x] Implement the RFC (https://github.com/rust-lang/rust/pull/87688)
- [ ] Adjust documentation: reference PR at https://github.com/rust-lang/reference/pull/1156
- [ ] Stabilization PR: https://github.com/rust-lang/rust/pull/93628

[doc-guide]: https://rustc-dev-guide.rust-lang.org/stabilization_guide.html#documentation-prs
[stabilization-guide]: https://rustc-dev-guide.rust-lang.org/stabilization_guide.html#stabilization-pr

### Unresolved Questions

- [x] **Readability in practice**: Will `let ... else { ... };` be clear enough to humans in practical code, or will some introducer syntax be desirable?
- [x] **Conflicts with if-let-chains**: Does this conflict too much with the if-let-chains RFC or vice-versa? Neither this feature nor that feature should be stabilized without considering the other.
- [x] **Amount of special cases**: Are there too many special-case interactions with other features?
- [x] **Grammar clarity**: Does the grammar need to be clarified? This RFC has some slightly unusual grammar requirements.

### Implementation history

- https://github.com/rust-lang/rust/pull/87688 implemented let-else.
- https://github.com/rust-lang/rust/pull/89841 adds support for type annotations in let-else (`let x: Type = ... else { ... }`)

**Comment by inquisitivecrystal** [comment]
Minor meta notes: the link to the RFC is broken and the F-let-else label is the wrong color. :)

This comment should probably be collapsed once they've been fixed.

**Comment by LeSeulArtichaut** [comment]
Is there already someone who wants to implement the RFC? Otherwise I could maybe take a stab at it

**Comment by camsteffen** [comment]
I volunteered on Zulip to implement a while ago and I have a start on it. @Fishrock123 are you still okay with me taking it since you were also interested?

**Comment by Fishrock123** [comment]
I'm interested but I'm clueless and pretty busy. I'd say you should just take it.

**Comment by matklad** [comment]
An interesting example which I didn't see come up in the RFC text/thread (quick search for `mut`). In today's Rust, `let-else` won't work for a case of rebinding variables. An example here is computing a path in a graph after DFS returns a map of parents: 

```rust
let parents: HashMap<u32, u32> = dfs();
let mut curr: u32 = 0;
let mut path: Vec<u32> = Vec::new();
loop {
    path.push(curr);
    let Some(&curr) = parents.get(curr) else { break }
}
```

Here, `let` won't re-bind curr, it'll shadow it instead (I think we have plans for allowing re-binding behavior in such cases, not sure what's the status of that). 

Just as an observation, Kotlin-style elvis operator would avoid this issue, as it doesn't require a pattern:

```rust
let parents: HashMap<u32, u32> = dfs();
let mut curr: u32 = 0;
let mut path: Vec<u32> = Vec::new();
loop {
    path.push(curr);
    curr = *(parents.get(curr) ?: break);
}

```

**Comment by joshtriplett** [comment]
On Wed, Aug 18, 2021 at 04:35:50AM -0700, Aleksey Kladov wrote:
> An interesting example which I didn't see come up in the RFC text/thread (quick search for `mut`). In today's Rust, `let-else` won't work for a case of rebinding variables. An example here is computing a path in a graph after DFS returns a map of parents: 
> 
> ```rust
> let parents: HashMap<u32, u32> = dfs();
> let mut curr: u32 = 0;
> let mut path: Vec<u32> = Vec::new();
> loop {
>     path.push(curr);
>     let Some(&curr) = parents.get(curr) else { break }
> }
> ```
> 
> Here, `let` won't re-bind curr, it'll shadow it instead

I wouldn't *ever* expect `let` to change what an existing location
refers to, only to make a new binding and shadow any previous binding.

**Comment by matklad** [comment]
Ah, indeed, I am confused, I was thinking about https://github.com/rust-lang/rust/issues/71126, which is a different feature, I'll edit my comment.

**Comment by matklad** [comment]
Hypothetically, we can allow the following (else without let):

```rust
let parents: HashMap<u32, u32> = dfs();
let mut curr: u32 = 0;
let mut path: Vec<u32> = Vec::new();
loop {
    path.push(curr);
    Some(&curr) = parents.get(curr) else { break }
}
```

This would be a natural combination of let-else and #71126. While #71126 is still unstable, it might be worth is to think about the interactions between the two. Will the above example just work, or are there any problems with it?

**Comment by camsteffen** [comment]
I also [mentioned](https://github.com/rust-lang/rfcs/pull/3137#issuecomment-866801443) this in the RFC. Probably no technical limitation but just readability concerns. I think we can safely wait until let-else and destructuring-assignment have matured before expecting them to procreate (forgive me).

**Comment by Fishrock123** [comment]
Ah, I seem to have missed that think about `destructuring_assignment`, yes that seems like a future possibility, with the above mentioned caveats.

**Comment by Fishrock123** [comment]
@joshtriplett https://github.com/rust-lang/rust/pull/87688 has been merged, can you tick off "implement the RFC"?

**Comment by Fishrock123** [comment]
@camsteffen @cjgillot did the implementation answer the "Grammar clarity" question at all?

**Comment by camsteffen** [comment]
> @camsteffen @cjgillot did the implementation answer the "Grammar clarity" question at all?

See [`expr_trailing_brace`](https://github.com/rust-lang/rust/blob/ad3407f482d29f26c67700974d86f89b3f2ab993/compiler/rustc_ast/src/util/classify.rs#L28-L52) - I think the "slightly unusual grammar requirements" are contained there.

**Comment by Fishrock123** [comment]
My understanding of the state of implementation is:
- Working on nightly under a flag (I actually have yet to try it! I've been busy)
- Flagged for compiler perf investigation https://github.com/rust-lang/rust/pull/87688#issuecomment-915318420

**Comment by est31** [comment]
I've tried out the feature in the compiler to help gather some practical experience with it, and maybe help getting it stabilized, see PR #89933. I first want to say that it is an amazing feature and indeed beautifies code a lot. I've been a great fan of it before, and my appreciation for it has only increased.

I've found two issues, both which I think should be resolved before stabilization:

#### `else if` unit tests and error message

The RFC has no mention of supporting `else if` instead of just bare `else`. The implementation also doesn't support it. Maybe one can later add support for it if all branches diverges and there is a final else, but it's perfectly fine to not have it for the moment and ship an MVP. However, before the let_else feature is stabilized, I think there need to be unit tests in the compiler's test suite ensuring that `else if` is indeed disallowed. I grepped for "else if" in the `src/test/ui/let-else` directory and couldn't find anything.

Also I think the error message for `let ... else if` is a bit unfriendly. It says "expected `{`, found keyword `if`". This is not wrong, but IMO it should specifically mention that `else if` is not supported. I do want to call out the very helpful suggestion though to wrap the `if` in `{}`. Maybe there should be a rustfix unit test for it?

#### bindings vs values

This is more an issue with the design of `let ... else` instead of the implementation. The problem is about the confusability of bindings with values. Consider code like:

```Rust
let Hello(Foo, bar) = val else { return };
```

There is no insurance that `Foo` is a value, and `bar` is indeed a new binding. This same problem exists everywhere where patterns exist which accept bindings, so in `if let`, `match`, etc. Generally, there are two things which alleviate the situation:

* The limited scope. Both `if let` and match clauses have a clearly ending scope with a `{}` block.
* The convention. Per naming convention, `Foo` indicates a type or the constructor for it, so one can see that it is a value, and `bar` indicates a value.

With `let ... else` the limited scope disappears, because suddenly all bindings live on for the entirety of the surrounding block. The only thing keeping code readable is the convention, but it is only enforced via default-on lints that can be turned off, and sometimes are. I'm not sure this is enough, and think the initially stabilized implementation should disallow patterns that contain values that might be mistaken for bindings, at least initially until there has been a decision about what to do about them.

Generally, `let ... else` makes it harder to recognize when a new binding is added to the context, compared to `let ident = if let ...`, because the binding can be quite hidden in the pattern. But removing that text is the point of the sugar :). I think that the strongest manifestation of the issue is in the binding vs value confusion outlined above, but otherwise it's worth the reduction in verbosity.

**Comment by est31** [comment]
The lint is a nice idea, and is definitely an option. One problem though is the question of actionability. If you don't choose the name for a type or a constant, but you want to use it, can you do something about it?

Another problem to consider is internationalization. Many scripts don't have concepts of large vs small letters. What should be done about them?

**Comment by digama0** [comment]
> The lint is a nice idea, and is definitely an option. One problem though is the question of actionability. If you don't choose the name for a type or a constant, but you want to use it, can you do something about it?

I think the main time you will be interacting with this lint is when you are using a library that requires nonstandard names for some reason or another (e.g. `libc`). Like I said, it's totally okay to just allow the lint whenever it comes up, because the point is to get that `#[allow]` directive close to the binding so that readers know to be wary. Alternatively you can just get peeved about it and `#[allow]` it at the top level, and maybe that's the right solution for a project that completely diverges from rust casing convention.

> Another problem to consider is internationalization. Many scripts don't have concepts of large vs small letters. What should be done about them?

The ideal solution here is for the lint to fire whenever the name of a type or constant has a name which matches the local convention for variable names, whatever that is. As far as I know this isn't really configurable, so the lint will probably not be useful when everything is written in a language for which rust doesn't know the convention.

**Comment by cormacrelf** [comment]
> @est31: For full transparency, I've held back some few changes from #89933 that created borrow checker errors. I believe that #89841 will fix these issues

Can we see them anyway? Drop them in a draft PR perhaps, it would be great to try it. In general, this being a new syntax, we need it.

Some of the codepaths in let-else generally are pretty thoroughly tested because `if let` bindings desugar and use this code as well, so even just compiling std is a lot. There is practically zero for the optional type annotation in combination with a refutable type annotation + inference + deref coercion (as that is not something any other syntax involves afaik), so if anyone wants to do more focused testing then that is where to direct your energy. Current impl desugars that part too, but it will need testing if 89841 or similar manages to supplant it.

**Comment by est31** [comment]
@cormacrelf sure you can see the changes. I've pushed the changes to a [separate branch](https://github.com/rust-lang/rust/compare/master...est31:let_else_borrow).

**Comment by oconnor663** [comment]
I was playing with this today, and the line I came up with (in the middle of a guessing game parse loop) was:

```rust
let guess: i64 = line.expect("io error").parse() else { continue };
```

I thought it might be kind of nice if I was allowed to omit the curly braces, like this:

```rust
let guess: i64 = line.expect("io error").parse() else continue;
```

Is that something folks have discussed before? I realize it's dangerously close to "braces optional after if-statements", which we know is a bad idea from decades of experience with C and C++. But maybe there's room for a special case here related to diverging keywords, that might apply both to let-else and also to regular if-statements? Something like:

```rust
if my_condition() {
    do_something();
} else return false;
```

Anyway, could be a terrible idea. I'm not sure what the consequences would be for parsing nested statements. But the slightly unfortunate curly braces around my `continue` at the top got me thinking about it.

**Comment by joshtriplett** [comment]
Based on @est31's trial-run PR in the compiler, I'm marking the unresolved question "Readability in practice" as resolved.

**Comment by nikomatsakis** [comment]
> With let ... else the limited scope disappears, because suddenly all bindings live on for the entirety of the surrounding block. The only thing keeping code readable is the convention, but it is only enforced via default-on lints that can be turned off, and sometimes are. I'm not sure this is enough, and think the initially stabilized implementation should disallow patterns that contain values that might be mistaken for bindings, at least initially until there has been a decision about what to do about them.

Like others, I feel like this is a pre-existing danger that is not made particularly worse by let else, though I confess I've not tried using it in practice. The argument for the "extended scope" having an impact is that the use site might be farther from the definition site?

I do feel that *not* supporting unit struct patterns would be a significant missing piece. Perhaps I'm weird, but I do often use "sentinel" values for errors and things, so being able to write

```rust
let Ok((SomeSentinel, value)) = foo() else { return; }
```

is something I care about (admittedly, the sentinels occur more often in the `Err` case). It seems to me like having a subset of patterns that work would be rather surprising.

**Comment by est31** [comment]
> The argument for the "extended scope" having an impact is that the use site might be farther from the definition site?

Yeah the potential use sites at least.

But I agree, it's already an existing issue. I don't think it should block the feature, and a discussion has been had (which was my main wish when I brought it up, just didn't want stabilization to happen without awareness of the issue). So I'd say it's resolved.

With #89841 merged, I'd say the borrow problems are gone now, although I didn't have the chance to try it out. I wonder if it might be a good idea to make a list of blockers for stabilization.

The feature is also very popular per this unscientific account: https://www.reddit.com/r/rust/comments/rdsf8p/how_have_i_lived_without_letelse/

**Comment by a1phyr** [comment]
I haven't found any discussion about how neatly this feature works with `try` blocks:

```rust
let Ok(res) = (try {
    let x = get_x()?;
    let y = get_y()?;
    combine(x, y)?
}) else {
    eprintln!("An error occurred !");
    return;
}
```

However today parenthesis are required around the `try` block, which makes it feel a weird and second-class. Would it be possible to allow eluding them in this special case ?

Or would it be too confusing ?

**Comment by joshtriplett** [comment]
@a1phyr We haven't yet committed to never supporting `try { ... } else { ... }` as valid syntax, and if that became valid syntax then `let pat = try { ... } else { ... }` would be ambiguous.

**Comment by Fishrock123** [comment]
I'd like to see it stabilized but I can't offer more than a surface opinion, since I haven't really been able to use it.

I never really got to try it at my former employer because we were pretty set on using stable (or perhaps I was not adventurous enough, regardless, I was pretty busy after the summer).

I'd been taking some time off, but now that I'm coming back to things I have a similar issue: nothing that I work on (mostly open source libraries) is able to use nightly, at least not without compile time feature gates.

I think a lot of this boils down to it being convenience rather than a hard "can't do it without that", for better or worse.

**Comment by camsteffen** [comment]
I think we're blocked on style guide and rustfmt support. I for one am confident with the functionality.

**Comment by calebcartwright** [comment]
> I think we're blocked on style guide and rustfmt support. I for one am confident with the functionality.

FWIW stabilization has never hinged on anything rustfmt related as far as I know, so I certainly wouldn't view that as a prerequisite.

**Comment by joshtriplett** [comment]
I agree that this is ready to stabilize, based on the feedback we've already gotten for how useful it is and how well it works.

**Comment by Fishrock123** [comment]
Regarding the still unresolved questions:

- **Conflicts with if-let-chains**: Are we saying it is let-if-chains' responsibility to figure out any interactions, or is there something more we need to do there?
- **Amount of special cases**: This doesn't seem to have been an issue?
- **Grammar clarity**: @camsteffen Have there been any significant changes in the implementation grammar that should be reflected in an RFC amendment?

**Comment by camsteffen** [comment]
> * **Grammar clarity**: @camsteffen Have there been any significant changes in the implementation grammar that should be reflected in an RFC amendment?

Not that I know of.

**Comment by est31** [comment]
Given the positive voices towards stabilization earlier in the thread, I've filed PR for such: #93628 (Currently a draft as it needs FCP). I've also filed a [PR for the reference](https://github.com/rust-lang/reference/pull/1156).

**Comment by cormacrelf** [comment]
> Grammar clarity: @camsteffen Have there been any significant changes in the implementation grammar that should be reflected in an RFC amendment?

For clarity, while the summary of my PR in this issue's header says I "added support for type annotations", they were supported before, the user-facing grammar has not been touched since Cam's work. My PR title was a bit misleading in this regard, should have said "re-implement" or "new desugaring".

**Comment by cormacrelf** [comment]
Oh, but it is worth surfacing from the implementation work that the changes under the hood have made it very easy to permit type annotations on if-let, bringing them in line with let-else. That is not implemented here and would require a new RFC, but it's very low-hanging fruit.

The HIR for if-let can now represent and type-check type annotations, the only thing missing is support from the parser. The case for it in short is `if let Ok(x): Result<i32, _> = num.try_into() { ... }`, and that it is a bit weird that you can do that with let-else but not if-let.

**Comment by LHolten** [comment]
My bad if this has been suggested before;

Instead of requiring the "let else block" to diverge, you could make it work the same as the `else` case of an `if let` statement.
This implies a straigtforward desugar:
```
let Some(val) = my_option else {
    // snippet 1
};
// snippet 2
```
always desugars to
```
if let Some(val) = my_option {
    // snippet 2
} else {
    // snippet 1
}
```
As a result we would have implicit return for `let else` statements outside loops and implicit continue for `let else` statements inside loops.
```
let Some(_) = my_option else {Err("wut")}; // implicit return
for item in option_list {
    let Some(val) = item else {}; // implicit continue
    println!("{val}");
}
```
Advantage of this is the simple desugar and shorter code.
Disadvantage might be that meaning depends more on context.

**Comment by andylizi** [comment]
> you could make it work the same as the `else` case of an `if let` statement.

This would probably cause endless confusion…… No one could possibly guess that `snippet 2` won't be executed, as this behavior differs from literally every programming language in existence.

**Comment by andylizi** [comment]
Yeah, "every language in existence" is probably a bit of an exaggeration. I was more referring to the behavior of "code blocks". While we're reading code and we see a closing brace, we wouldn't expect it to implicitly diverge. For example:
```rust
{
    let foo = {

        // some long lines of code, can't be displayed in one screen

    };  // <--- this won't diverge
}
```
```rust
{
    let Some(foo) = ... else {

        // some long lines of code, can't be displayed in one screen

    };  // <--- this will diverge???
}
```
It really does go against one of the first instinct we got for curly-braced language.

And while I know little about Lean or functional languages in general, I managed to find some examples using [this grep.app search](https://grep.app/search?q=let%20%5B%5E%7B%5Cn%5D%2B%20%5C%7C%20&regexp=true&case=true&filter[lang][0]=Lean):
```lean
let some n := ln.natLit? | throwError "fail"
def ident : ParserM Name := do let ⟨_, VMCall.ident n⟩ ← next | failure; pure n
let some (_, lhs, rhs) := eq.eq? | throwError "not an equality {eq}"
let some (_, lhs', rhs') := ty'.eq? | return none
let ([x], ps) := processConstructor pat.ref 1 pat.asTuple | panic! "rcases"
let Expr.const unsafeFn unsafeLvls .. := t.getAppFn | unreachable!
```
They seems to convey the meaning of "else we diverge" pretty clearly to me. And it's not just functional languages either. PHP for example:
```php
mysql_connect($host, $user, $pass) or die("could not connect");
```

**Comment by digama0** [comment]
> Yeah, "every language in existence" is probably a bit of an exaggeration. I was more referring to the behavior of "code blocks". While we're reading code and we see a closing brace, we wouldn't expect it to implicitly diverge.

It's not implicitly diverging, it's going to the end of the block. Saying that it "implicitly diverges" is the same as arguing that a `loop `"implicitly diverges" when you reach the end of the body and go back to the start, or an `if` "implicitly diverges" when you reach the end of the then block and skip the rest of the `else if ... else` stuff. It's a jump to a lexically well defined location (the end of the enclosing block).

That is to say, it's a reasonable choice for a language to make. I'm not fussed about Rust making the call that this is too confusing and we should just ban it, though.

> And while I know little about Lean or functional languages in general, I managed to find some examples using [this grep.app search](https://grep.app/search?q=let%20%5B%5E%7B%5Cn%5D%2B%20%5C%7C%20&regexp=true&case=true&filter%5Blang%5D%5B0%5D=Lean):
> 
> ```lean
> let some n := ln.natLit? | throwError "fail"
> def ident : ParserM Name := do let ⟨_, VMCall.ident n⟩ ← next | failure; pure n
> let some (_, lhs, rhs) := eq.eq? | throwError "not an equality {eq}"
> let some (_, lhs', rhs') := ty'.eq? | return none
> let ([x], ps) := processConstructor pat.ref 1 pat.asTuple | panic! "rcases"
> let Expr.const unsafeFn unsafeLvls .. := t.getAppFn | unreachable!
> ```
> 
> They seems to convey the meaning of "else we diverge" pretty clearly to me.

Well yes, let-else is all about "else yeet" but the question is "yeet where?" Of these examples, all of them are (roughly) actual divergence in the rust sense except for the second one, which calls a `failure` function which does not do any funny control flow and then passes `n` to the enclosing block (in this case it amounts to a return anyway, though).

If you always use a diverging expression, the question never comes up, but you are syntactically capable of putting something that has a different type in that slot, and it's a reasonable question to ask "where does that value go?" and I would say that @LHolten 's suggestion is the most logically coherent choice. (Or you can just [unask the question](https://en.wikipedia.org/wiki/Mu_(negative)) like rust does.)

**Comment by Fishrock123** [comment]
```rust
{
    let Some(foo) = ... else {

        // some long lines of code, can't be displayed in one screen

    };  // <--- this will diverge???
}
```

This example is not good. Code will not look like this in practice.

“Must diverge” in let-else is enforced in a specific way: it must return `!` (the never type).

This explicitly means that a let-else block is almost certainly going to end with one of a few options: `return`, `break`, `continue`, `panic!` (or another macro which invokes `panic!`). Any other options are obscure and unlikely to be encountered, I think, though more could be added to the language in the future.

**Comment by itsVyasi** [comment]
> Instead of requiring the "let else block" to diverge, you could make it work the same as the `else` case of an `if let` statement. This implies a straigtforward desugar:
> 
> ```
> let Some(val) = my_option else {
>     // snippet 1
> };
> // snippet 2
> ```
> 
> always desugars to
> 
> ```
> if let Some(val) = my_option {
>     // snippet 2
> } else {
>     // snippet 1
> }
> ```

I'm not sure that is a simple desugar, since snippet 2 isn't guaranteed to be the entire following block. I.e.
```
let Some(val) = my_option else {
    panic!();
};
process(val);
// do other stuff
```
would need to be
```
if let Some(val) = my_option {
    process(val);
} else {
    panic!();
}
// do other stuff
```

I think the current design is more intuitive, especially since the `else` in let-else is buried in the `let` statement and could be overlooked, whereas other divergent statements are visible at the head of the line: `return`, `loop`, `while true`, `panic!()`, etc.

The fix there I think would be 
```
let Some(val) = my_option else {
    // snippet 1
} { 
    // snippet 2
} 
```
which is strikingly close to the `if let val = EXPR else { ... } { ... }` anti-pattern, since it reverts the design goal of let-else to move error code close to the error checking conditional.

**Comment by LHolten** [comment]
> I'm not sure that is a simple desugar, since snippet 2 isn't guaranteed to be the entire following block. I.e.
> 
> ```
> let Some(val) = my_option else {
>     // snippet 1
> };
> // snippet 2
> // snippet 3
> ```
> 
> could need to be
> 
> ```
> if let Some(val) = my_option {
>     // snippet 2
> } else {
>     // snippet 1
> }
> // snippet 3
> ```
I did intend for snippet 2 to be the entire block. So your example would **always** desugar to this:
```
if let Some(val) = my_option {
    // snippet 2
    // snippet 3
} else {
    // snippet 1
}
```
It needs to be clear that desugaring is a one way operation. In this case it is possible to turn every `let else` into an `if let else`, but it is not possible to turn every `if let else` into a `let else`.

This is exactly the same as the current design. 
Actually, the current design with divergence check can be implemented with the exact same desugar. The only difference is that it needs to check for divergence of the generated `else` branch. Removing the check would then yield "implicit divergence" as was my proposal.

Btw, everything that compiles with the current design would compile exactly the same with "implicit divergence".
So we could stabilize the current design first and then add support for "implicit divergence" by removing the check for divergence.

**Comment by Eugeny** [comment]
IMO this breaks the principle of _locality_, where the effects to control flow are represented by local curly braces.

I.e. here I _know_ that the `foo` call is a static part of the function body since it's in the function body's block.

```
fn x() {
    [...]
    foo();
} // <- here
```

and here it's affected by some condition since it's in the `if` block:

```
fn x() {
  if ... {
    [..]
    foo();
  } // <- here
  [..]
}
```

But with your desugaring, the first snippet immediately becomes ambiguous when reading:


```
fn x() {
  if ... {
    [..]
    foo();
  } // <- sure, it's in an `if`, but was there some hard to see `let` above this that could have affected the control flow at this point?
  [..]
}
```

**Comment by LHolten** [comment]
> IMO this breaks the principle of locality, where the effects to control flow are represented by local curly braces.

You could say the same for the question mark operator `?`.
It is very small, it is not located at the start of a line and it affects control flow in the same way.

**Comment by Eugeny** [comment]
I've done some self-reflection and I think the actual thing that bugs me is that it's introducing a brand new control flow concept.

The current control-flow-breaking "early return" actions we have are:

* Exit function: `return`, `panic` and `?`
* Exit loop body: `break`

The alternate let-else desugaring introduces the new concept of "exiting the current code block but not the function".

If the flexibility of not having to exit the function is a goal, it might be worth to look at adding "exit scopes", similarly how the try-blocks RFC is doing for the `?` operator.

**Comment by LHolten** [comment]
> The alternate let-else desugaring introduces the new concept of "exiting the current code block but not the function".

I agree with your analysis. The reason I think this should be the default for a `let else` binding, is because it is precisely the current code block that requires the variable to be bound. Execution can continue after the current block and I think it should.

Since it probably requires a lot of testing to figure out if I am right that this is asked for. Why not continue stabilising the current design while making sure not to sabotage any future implementation of the more relaxed alternative desugar?

**Comment by digama0** [comment]
@joshtriplett Could the OP be updated to point to the reference PR (rust-lang/reference#1156) and the stabilization PR (#93628)? They both seem to be waiting on this issue to be merged first.

**Comment by Liamolucko** [comment]
I took a look at the HIR and it looks like this:

```rs
let Some(num1) = MUTEX.lock().unwrap().func() else {
    unreachable!()
};
let Some(num2) = MUTEX.lock().unwrap().func() else {
    unreachable!()
};
// ...
```

gets desugared to this:

```rs
if let Some(num1) = MUTEX.lock().unwrap().func() {
    if let Some(num2) = MUTEX.lock().unwrap().func() {
        // ...
    } else {
        unreachable!();
    }
} else {
    unreachable!();
}
```

Because the code after the `let ... else` gets put inside of the `if let`, the guard hasn't yet been dropped and it deadlocks.

That's not what the RFC originally proposed. It proposed (and I assumed) that it would get expanded to this:

```rs
let num1 = match MUTEX.lock().unwrap().func() {
    Some(num1) => num1,
    _ => unreachable!(),
};

let num2 = match MUTEX.lock().unwrap().func() {
    Some(num2) => num2,
    _ => unreachable!(),
};

// ...
```

That version doesn't deadlock, since the guard gets dropped after the first `match`.

It looks like this change happened when first implementing `let ... else`: https://github.com/rust-lang/rust/pull/87688#discussion_r681278414

**Comment by MatrixDev** [comment]
Personally I'm using following approach until let-else is stabilised, but match version is basically the same:
```rust
let value = if let Some(e) = &option { e } else {
    unreachable!();
};
```

**Comment by hardskulls** [comment]
Why not extend `?` operator? Some people already mentioned it here. I mean it's purpose is to get the wanted type and unwrap it or to return unwanted if something goes wrong. 
It could be from something simple like:

```rust
let msg = Some("Hi!");
let num: i32 = msg
                 ?:{ Some(_) | Err("No text here!")}
                 .parse()
                 ?:{ _ else {
                         ...  ;
                         return Err("Text is not a number!" 
                       }
                 };
```

To a more complex syntax like:

```rust
let var_name = EXPR?: WANTED_TYPE 
                   else break TYPE_1 | TYPE_2 | TYPE_3 // If this is in a loop, for example.
                   else return TYPE_4(val) if val == ... | TYPE_5
                   else/default { 
                         ... ;
                         return/break/(!) ... 
                   };

```

Where all types are variants of single enum and which would be read as "If the wanted type is not present, break the loop with types 1-3, if they're not present, return the function with types 4-5, and if everything above fails execute a block returning `!` type". The only problem I see here is method calls, which need to be separated from `?: ...` with braces.

**Comment by joshtriplett** [comment]
It looks like https://github.com/rust-lang/rust/issues/98672 has now been fixed. Is that the last blocker here? Should we r+ the stabilization PR and the documentation PR at this point, or does more need to happen before that?

**Comment by est31** [comment]
There is this comment by @pnkfelix where he asks for more tests: https://github.com/rust-lang/rust/pull/93628#issuecomment-1055738523

#98574 has added a bunch of tests but only one that tests for the drop order. The test case it added can be expanded. It should be easy to make a PR for it. I want to do so in the coming days.

Outside of that I don't know of any blockers.

**Comment by camsteffen** [comment]
Looks like it still does not drop before the `else` block so I reopened #98672. And also I think there needs to be a formal decision on *whether* temps should drop before the `else` block.

**Comment by joshtriplett** [comment]
I agree: temporaries should drop before the else block.

**Comment by hikari-no-yume** [comment]
If temporaries don't drop, this would make let-else a lot less useful in some circumstances. Consider a scenario using RefCell:

```rust
let Some(copied_val) = cell.borrow().val else { panic!() };
may_mutate_cell(&cell, copied_val);
```

When `may_mutate_cell` attempts a mutable borrow of the cell, a panic will happen. I think that's surprising and unhelpful.

**Comment by Fishrock123** [comment]
I’m not sure why the above comment was edited out, it seems like a mistake. It had the following comment:

>> If temporaries don't drop, this would make let-else a lot less useful in some circumstances. Consider a (heavily simplified) scenario using RefCell:
>
> ```rust
> let Some(copied_val) = cell.borrow().val else { panic!() };
> may_mutate_cell(&cell, copied_val);
> ```
> When `may_mutate_cell` attempts a mutable borrow of the cell, a panic will happen. I think that's surprising and unhelpful.

**Comment by hikari-no-yume** [comment]
Yes sorry, I edited it out because I realised others were currently talking about dropping after `else` not dropping in general. I suppose something similar can apply to the `else` though, my example is just not immediately helpful there.

**Comment by est31** [comment]
I've made a PR for the drop order tests: #99291

It's sadly indeed all over the place:
```
v is 0 (match), construct is let else
  vanilla:      drop,body,
  &:            body,drop,
  &mut:         body,drop,
  tuple:        drop,body,
  array:        drop,body,
  nested:       drop,inner,body,
  ref &:        body,drop,
  ref mut &mut: body,drop,
```
Compare this to `if let`:
```
v is 0 (match), construct is if let
  vanilla:      body,drop,
  &:            body,drop,
  &mut:         body,drop,
  tuple:        body,drop,
  array:        body,drop,
  nested:       body,drop,inner,
  ref &:        body,drop,
  ref mut &mut: body,drop,
```

**Comment by est31** [comment]
From the added tests in #99291 I made a summary of the temporary drop order behaviour of the current implementation, comparing it to existing constructs:

| construct | order on match | order on mismatch |
|-|-|-|
| if let | body, drop | else, drop |
| match | body, drop | else, drop |
| if | drop, body | drop, else |
| let else most cases | drop, body | else, drop |
| let else & | body, drop | else, drop |

if let chains behave like `if let` for temporaries in a `let` construct, and like `if` for temporaies in a bool typed expression. Personally I think this choice is reasonable, but maybe the question should be discussed for them as well? cc #94927

So `let else` behaves like if let and match on mismatch. When the pattern matches, drop order depends on the pattern. If there is a `&` or a `&mut`, e.g. in `let &CustomDrop(_) = &CustomDrop(foo) else {...}`, then it post-drops like if let and match. Otherwise, it pre-drops like `if`.

If you want to maximize consistency, and this was an important goal for let else desugaring before, you should probably make it consistent to if let and match. But it's also possible to change the implementation to pre-drop as often as possible (always doable in the `else` case, with & in the match case you probably still need it as there is no move happening).

**Comment by joshtriplett** [comment]
I *think* the behavior on success is now what we want. I think it's just the behavior for the else case that we want to change: I think we should consistently always drop before the else.

**Comment by wmstack** [comment]
What do you think of allowing non `!` types for the else branch of the statements? The idea is this, that a let statement creates new bindings for variables. In any situation, let statements don't have the job of initializing those variables. Normal let statements can also create uninitialized variables, and refutable patterns just embody *possibly uninitialized values*. Rust by default prevents you from using uninitialized variables. If there is a possibility that the pattern is refutable, rust will warn us that the variable is not necessarily initialized so we can't access it after-wards. Then we go back and add an `else` branch:

```rust
let Some(x) = option else {
  /* handle case of possibly uninitialized variable */
  x = 42;
}

/* variable is guaranteed to be initialized. */
```

We already can have let statements that don't initialize their bindings, so I don't see the problem of extending irrefutabe patterns to refutable patterns in `let` statements this way.

**Comment by Fishrock123** [comment]
@wmstack Please see the [“Future Possibilities: Fallback Assignment”](https://github.com/rust-lang/rfcs/blob/master/text/3137-let-else.md#fall-back-assignment) section of the RFC.

The RFC leaves this open as a possible addition by using some additional syntax. See [the “`else`-clock fall-back assignment” rationale section](https://github.com/rust-lang/rfcs/blob/master/text/3137-let-else.md#else-block-fall-back-assignment) for why.

**Comment by kevincox** [comment]
Strong agree. Let's start with the simple case and we can consider options like that in a new RFC with some real-world experience under our belt.

**Comment by est31** [comment]
In *theory* there are some scenarios where we could drop earlier. E.g. in this one:


```Rust
#![feature(let_else)]

struct Droppy(u32);

impl Drop for Droppy {
    fn drop(&mut self) {
        println!("I dropped");
    }
}

#[allow(irrefutable_let_patterns)]
fn main() {
    {
        let &Droppy(v) = &Droppy(32) else { 
            todo!()
        };
        println!("v is {v}");
        println!("Drop will happen soon");
    }
    println!("--- Next ---");
    {
        let &Droppy(_) = &Droppy(32) else { 
            todo!()
        };
        println!("I should have dropped");
    }
}
```

It prints:
```
v is 32
Drop will happen soon
I dropped
--- Next ---
I should have dropped
I dropped
```

The first example makes sense: `v` is not moved out of `Droppy` because we only have it by reference. In the second example, however, no binding is created, we only do pattern matching on Droppy. Basically we could replace it with a use of `matches!()`.

Note that this *is* similar however to what we do with irrefutable patterns already. They drop immediately when it's a by-val pattern match but live on if it's a by-ref pattern match, even if no bindings are being created.

```Rust
    {
        let Droppy(_) = Droppy(32);
        println!("After the statement");
    }
    println!("--- Next ---");
    {
        let &Droppy(_) = &Droppy(32);
        println!("After the statement");
    }
```
Prints:
```
I dropped
After the statement
--- Next ---
After the statement
I dropped
```

Should I open a new issue for discussion?

**Comment by est31** [comment]
> In theory there are some scenarios where we could drop earlier. 

Answering myself: I think [this](https://github.com/rust-lang/rust/pull/93628#issuecomment-1047140316) comment by @joshtriplett resolves this: We should just do whatever `let` is doing. As the current behaviour matches `let`, I think we can stabilize it this way (although tests for it would be good, reminder that I add them to #99291).

**Comment by oxalica** [comment]
The current proposal-ed syntax accepts no pattern in the `else` branch. This makes it unhelpful with another common pattern,
```rust
let val = match ... {
    Ok(val) => val,
    Err(e) => { /*use e here!*/ return; } // Or `panic`, `continue`, whatever.
};
```
We may workaround by `let Ok(val) = (...).map_err(|e| /* use e */) else { return; }` but it is 1. ugly, 2. fail to compile if both `Ok` and `Err` branches take ownership of some variables.

Note that [Idris has a similar syntax](http://docs.idris-lang.org/en/latest/tutorial/interfaces.html?highlight=monad#pattern-matching-bind) which supports "else-arms" in case of failed let binding.
```idris
do
  Right x <- readNumber | Left y => ... -- <- Interrupts the normal control flow but can have own bindings.
  -- More actions...
```

This reminds me that general `let-else` should be a special `match` with one arm binding to current scope.
Could we do better in this case? Syntax like `let Ok(e) = ... else match { Err(e) => { ... } }` could work but seems a bit too complicated.

**Comment by Eugeny** [comment]
IMO at that point it's easier to just `let x = match ...` with a deviating Err branch
